File: solutions/3.Longest_Substring_Without_Repeating_Characters/longest-substring-without-repeating-characters.js
// Source : https://leetcode.com/problems/longest-substring-without-repeating-characters/
// Author : Zeming (Vantic) Zhou
// Date   : 2016-11-25

/**
 * @param {string} s
 * @return {number}
 */

var lengthOfLongestSubstring = function(s) {
	var	result = 0,
			start = 0,
			dict = {};
			
	for (var i = 0; i < s.length; i++) {
		var curChar = s[i];
		if (dict.hasOwnProperty(curChar) && dict[curChar] >= start) {
			start = dict[curChar] + 1;
		}
		dict[curChar] = i;
		result = Math.max(result, i - start + 1);
	}
	
	return result;
};

File: solutions/3.Longest_Substring_Without_Repeating_Characters/longest-substring-without-repeating-characters.py
# Source : https://leetcode.com/problems/longest-substring-without-repeating-characters/
# Author : Zeming (Vantic) Zhou
# Date   : 2016-11-25

class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        result = 0
        start = 0
        searched = {}
        
        for i, char in enumerate(s):
            if char in searched and searched[char] >= start:
                start = searched[char] + 1
            searched[char] = i
            result = max(result, i - start + 1)
        
        return result
